src/Tasks/FileIO/GetFileHash.cs

```diff
@@ -67,25 +75,27 @@ public override bool Execute()
                 return false;
             }
 
-            foreach (var file in Files)
+            Parallel.For(0, Files.Length, index =>
             {
+                var file = Files[index];
                 if (!FileSystems.Default.FileExists(file.ItemSpec))
                 {
                     Log.LogErrorWithCodeFromResources("FileHash.FileNotFound", file.ItemSpec);
                 }
-            }
+            });
 
             if (Log.HasLoggedErrors)
             {
                 return false;
             }
 
-            foreach (var file in Files)
+            Parallel.For(0, Files.Length, index =>
             {
-                var hash = ComputeHash(Algorithm, file.ItemSpec);
+                var file = Files[index];
+                var hash = ComputeHash(algorithmFactory, file.ItemSpec);
                 file.SetMetadata("FileHashAlgorithm", Algorithm);
```

Comment:
This call doesn't appear to be thread safe. You should do it under a lock or defer setting the metadata to after the parallel loop.

Reply:
I was thinking about duplicate items in the `Files` array but haven't succeeded in reproducing this. We may always wrap items with unique instances in which case this would indeed be safe. But it sure feels like an unnecessarily brittle invariant to rely on.

Reply:
It wouldn't be incorrect to take out duplicates right?

Reply:
Should be fairly easily to extract changing the array out if you think it's necessary. Let me know and I'll update accordingly.

Reply:
Team triage: Ladi has successfully scared Rainer about this. Please do extract the SetMetadata outside the Parallel.For.